source/Nevermore/Transient/DbCommandExtensions.cs

```diff
@@ -36,71 +36,57 @@ public static Task<int> ExecuteNonQueryWithRetryAsync(this DbCommand command, Re
                 var weOwnTheConnectionLifetime = EnsureValidConnection(command, connectionRetryPolicy);
                 try
                 {
-                    return await command.ExecuteNonQueryAsync();
+                    return await command.ExecuteNonQueryAsync(cancellationToken);
                 }
                 finally
                 {
                     if (weOwnTheConnectionLifetime && command.Connection != null && command.Connection.State == ConnectionState.Open)
-                        command.Connection.Close();
+                        await command.Connection.CloseAsync();
                 }
             });
         }
 
         public static DbDataReader ExecuteReaderWithRetry(this DbCommand command, RetryPolicy commandRetryPolicy, CommandBehavior behavior = CommandBehavior.Default, RetryPolicy connectionRetryPolicy = null, string operationName = "ExecuteReader")
         {
-            try
+            GuardConnectionIsNotNull(command);
+            var effectiveCommandRetryPolicy = (commandRetryPolicy ?? RetryPolicy.NoRetry).LoggingRetries(operationName);
+            return effectiveCommandRetryPolicy.ExecuteAction(() =>
             {
-                GuardConnectionIsNotNull(command);
-                var effectiveCommandRetryPolicy = (commandRetryPolicy ?? RetryPolicy.NoRetry).LoggingRetries(operationName);
-                return effectiveCommandRetryPolicy.ExecuteAction(() =>
+                var weOwnTheConnectionLifetime = EnsureValidConnection(command, connectionRetryPolicy);
+                try
                 {
-                    var weOwnTheConnectionLifetime = EnsureValidConnection(command, connectionRetryPolicy);
-                    try
-                    {
-                        return command.ExecuteReader(behavior);
-                    }
-                    catch (Exception)
-                    {
-                        if (weOwnTheConnectionLifetime && command.Connection != null &&
-                            command.Connection.State == ConnectionState.Open)
-                            command.Connection.Close();
-                        throw;
-                    }
-                });
-            }
-            catch (Exception ex)
-            {
-                throw new Exception($"Exception occurred while executing a reader for `{command.CommandText}`", ex);
-            }
```

Comment:
This exception handler was bad:tm:

It was catching unique constraint exception too early in the call chain and hiding the underlying sql error. The reason this changed in this PR is that previously it only threw the exception when the reader made the 1st call to `.Read()`, but the exception is now thrown here because the generated SQL is different and the exception is thrown at this point, rather than when read.